modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -79,43 +105,115 @@ object GlobalL0Service {
             )
         }
+
+      def pullGlobalSnapshotsFromMajority
+        : F[Either[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo), List[Hashed[GlobalIncrementalSnapshot]]]] = {
+        maybeMajorityPeerIds match {
+          case None => pullGlobalSnapshots
+          case Some(majorityPeerIds) =>
+            lastGlobalSnapshotStorage.getOrdinal.flatMap {
+              _.fold {
+                pullLatestSnapshotFromMajority.map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+              } { lastStoredOrdinal =>
+                for {
+                  snapshotData <- getLatestMajoritySnapshotData(majorityPeerIds)
+                  l0Peers <- globalL0ClusterStorage.getPeers
+                  lastOrdinal = snapshotData.ordinal
+                  majorityPeer <- getPeerInTheMajority(lastOrdinal, snapshotData.hash, l0Peers)
+                  pulled <- pullSnapshots(majorityPeer, lastStoredOrdinal.next, lastOrdinal)
+                } yield pulled.asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
+              }
+            }
+        }
+      }.handleErrorWith { e =>
+        logger.warn(e)(s"Failure pulling global snapshots from majority!") >>
+          Applicative[F].pure(
+            List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
+          )
+      }
+
+      private def pullSnapshots(
+        l0Peer: L0Peer,
+        nextOrdinal: SnapshotOrdinal,
+        lastOrdinal: SnapshotOrdinal
+      ): F[List[Hashed[GlobalIncrementalSnapshot]]] = {
+        val lastOrdinalCap = lastOrdinal.value.value
+          .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+
+        val ordinals = (nextOrdinal.value.value to lastOrdinalCap).toList
+          .map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o)))
+
+        (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
+          case (ordinal :: nextOrdinals, snapshots) =>
+            l0GlobalSnapshotClient
+              .get(ordinal)(l0Peer)
+              .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
+              .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=$ordinal")
+                  .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+              }
+
+          case (Nil, snapshots) =>
+            Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      // TODO should these majority functions move to MajorityPeerSelect.scala ?
```

Comment:
can you make this a conversation in the pull request? it's hard to track TODOs but it's easy to open a discussion there.
IMO, to move it to `MajorityPeerSelect` you'd have to write a new strategy for peer selection 🤔 I'd consider that as it may naturally follow-up what you've already down in majority peer select 🤔

Reply:
I would add the methods to `MajorityPeerSelect` but not change the existing ones. If there is an opportunity to refactor then I'd look at that in a later PR.

Reply:
Not sure if this is resolved but I'm fine with leaving this where it is in interest of expediency. Can be addressed as part of later PR

Reply:
Also, I raised the possibility of the `MajorityPeerSelect` class using a different strategy than that used for choosing the majority for snapshots, in which case they should remain separate.